### PR TITLE
Migration for checking guid uniqueness

### PR DESCRIPTION
**Pull request by d-m-u**
All the things that have the uuid mixin should validate that the guids generated are unique, I know it's a ridiculously stupid nit, and I know that we are so unlikely to ever find a case in which this migration is necessary (outside of ```dup``` which will be dealt with in the related PR) that I'm better off buying a lottery ticket for the three mill run today or whatever, but I wanted the satisfaction of having dealt with this so... 

TLDR blame Brandon. 

also, ```guidniqueness```?

Related:
https://github.com/ManageIQ/manageiq/pull/18492

There are two commits in this PR because providers currently have no validations on guids, nor are they required to have guids, and so since in the main repo PR we're adding the mixin to providers, which has a guid presence validation, this PR also checks to make sure that all providers do in fact have guids.

### CODE REVIEW

**Comment by d-m-u** [comment]
@miq-bot assign @bdunne

File: spec/migrations/20190226184255_check_guid_uniqueness_spec.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+       miq_region,
+       miq_server,
+       miq_widget,
+       miq_worker,
+       scan_item,
+       service,
+       service_template,
+       miq_set,
+       vm].each do |obj|
+        p obj.to_s
```
Comment (bdunne) [inline spec/migrations/20190226184255_check_guid_uniqueness_spec.rb]: Please remove debugging statements

File: spec/migrations/20190226184255_check_guid_uniqueness_spec.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        obj.create!.dup.save(:validate => false)
```
Comment (bdunne) [inline spec/migrations/20190226184255_check_guid_uniqueness_spec.rb]: What do you think of something more like this?

```ruby
guid = SecureRandom.uuid

obj_1 = obj.create!(:guid => guid)
obj_2 = obj.dup.save(:validate => false)

expect(obj_1.reload.guid).to eq(guid)
expect(obj_2.reload.guid).to eq(guid)

migrate

expect(obj_1.reload.guid).to eq(guid)
expect(obj_2.reload.guid).not_to eq(guid)
```

File: db/migrate/20190226184255_check_guid_uniqueness.rb [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    validates :guid, :uniqueness => true
+    self.inheritance_column = :_type_disabled # disable STI
+  end
+
+  def up
+    [AutomateWorkspace, ExtManagementSystem, Host, Job, LifecycleEvent, MiqAction, MiqAeWorkspace, MiqEventDefinition,
+     MiqPolicy, MiqRegion, MiqServer, MiqWidget, MiqWorker, ScanItem, Service, ServiceTemplate,
+     MiqSet, ScanItem, Vm].each do |thing|
+      say_with_time("Checking #{thing} for guid uniqueness") do
+        thing.where(:guid => thing.select(:guid).group(:guid).having("count(*) > 1").pluck(:guid))[0...-1].each { |obj| obj.guid = SecureRandom.uuid; obj.save! }
```
Comment (bdunne) [inline db/migrate/20190226184255_check_guid_uniqueness.rb]: I don't think you want the `pluck` in here, right?

Comment (bdunne) [reply db/migrate/20190226184255_check_guid_uniqueness.rb]: Also, the `[0...-1]` will leave out the last one.  I think we would actually want to leave the first record alone since it would be the original. (we may want to `order(:id)` to guarantee that)

Comment (bdunne) [reply db/migrate/20190226184255_check_guid_uniqueness.rb]: `.each { |obj| obj.guid = SecureRandom.uuid; obj.save! }` could be simplified to `.each { |obj| obj.update_attributes!(:guid => SecureRandom.uuid) }`

File: spec/migrations/20190226184255_check_guid_uniqueness_spec.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        expect(obj.where(:guid => obj.select(:guid).group(:guid).having("count(*) > 1").pluck(:guid)).count).to eq(2)
+        migrate
+        expect(obj.select(:guid).group(:guid).having("count(*) > 1").pluck(:guid).count).to eq(0)
+      end
+    end
+
+    it "automate workspace dup resets guid" do
+      automate_workspace.create!(:guid => SecureRandom.uuid).dup.save(:validate => false)
```
Comment (bdunne) [inline spec/migrations/20190226184255_check_guid_uniqueness_spec.rb]: I think this can be covered by the test above, right?

File: spec/migrations/20190226184255_check_guid_uniqueness_spec.rb (hunk 1 above)
Comment (d-m-u) [reply spec/migrations/20190226184255_check_guid_uniqueness_spec.rb]: I left this in because it's nice to see what object it's testing, is there a way of doing this that is better?

Comment (bdunne) [reply spec/migrations/20190226184255_check_guid_uniqueness_spec.rb]: Yes, the `[].each` can be moved up one level of nesting.

```ruby
migration_context :up do
  [things...].each do |test_subject|
    it "#{thing.name} guids are unique" do
      ...the test logic...
    end
  end
end
```

**Comment by bdunne** [comment]
@carbonin Can you take a look too? :eyes:

File: spec/migrations/20190226184255_check_guid_uniqueness_spec.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  let(:service) { migration_stub(:Service) }
+  let(:service_template) { migration_stub(:ServiceTemplate) }
+
+  migration_context :up do
+    %i(automate_workspace miq_ae_workspace miq_policy miq_region miq_widget service service_template).each do |obj|
+      it "#{obj} dup resets guid" do
+        guid = SecureRandom.uuid
+        klass = send(obj)
+        obj1 = klass.create!(:guid => guid)
+        obj2 = klass.new(:guid => guid)
```
Comment (carbonin) [inline spec/migrations/20190226184255_check_guid_uniqueness_spec.rb]: Can you also create a third object with a unique guid to ensure it isn't updated?

File: db/migrate/20190226184255_check_guid_uniqueness.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  class ServiceTemplate < ActiveRecord::Base
+    validates :guid, :uniqueness => true
+    self.inheritance_column = :_type_disabled # disable STI
+  end
+
+  def up
+    [AutomateWorkspace, MiqAeWorkspace, MiqPolicy, MiqRegion, MiqWidget, Service, ServiceTemplate].each do |thing|
+      say_with_time("Checking #{thing} for guid uniqueness") do
+        thing.where(:guid => thing.select(:guid).group(:guid).having("count(*) > 1").pluck(:guid)).order(:id).drop(1).each { |obj| obj.update_attributes!(:guid => SecureRandom.uuid) }
```
Comment (bdunne) [inline db/migrate/20190226184255_check_guid_uniqueness.rb]: Also, this should probably scope to `in_my_region` to avoid updating records from other regions

File: db/migrate/20190226184255_check_guid_uniqueness.rb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  class ServiceTemplate < ActiveRecord::Base
+    include ActiveRecord::IdRegions
+    validates :guid, :uniqueness => true
+    self.inheritance_column = :_type_disabled # disable STI
+  end
+
+  def up
+    [AutomateWorkspace, MiqAeWorkspace, MiqPolicy, MiqRegion, MiqWidget, Service, ServiceTemplate].each do |thing|
+      say_with_time("Checking #{thing} for guid uniqueness") do
+        thing.in_my_region.where(:guid => thing.select(:guid).group(:guid).having("count(*) > 1").pluck(:guid)).order(:id).drop(1).each { |obj| obj.update_attributes!(:guid => SecureRandom.uuid) }
```
Comment (bdunne) [inline db/migrate/20190226184255_check_guid_uniqueness.rb]: The `select` for finding the duplicates also needs the `in_my_region` scope, right?

File: spec/migrations/20190226184255_check_guid_uniqueness_spec.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        guid2 = obj3.guid
+
+        expect(obj1.reload.guid).to eq(guid)
+        expect(obj2.reload.guid).to eq(guid)
+
+        migrate
+
+        expect(obj1.reload.guid).to eq(guid)
+        expect(obj2.reload.guid).not_to eq(guid)
+        expect(obj3.guid).to eq(guid2)
```
Comment (carbonin) [inline spec/migrations/20190226184255_check_guid_uniqueness_spec.rb]: I think you need a reload here for this to be valid, right?

File: db/migrate/20190304192641_add_provider_guids.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+class AddProviderGuids < ActiveRecord::Migration[5.0]
+  class Provider < ActiveRecord::Base
+    include ActiveRecord::IdRegions
+    validates :guid, :uniqueness => true
+    self.inheritance_column = :_type_disabled # disable STI
+  end
+
+  def up
+    Provider.where(:guid => nil).each { |provider| provider.update_attributes!(:guid => SecureRandom.uuid) }
```
Comment (carbonin) [inline db/migrate/20190304192641_add_provider_guids.rb]: This looks like it's solving a different problem than uniqueness ... Can this be a separate PR or are they somehow related?

Comment (carbonin) [reply db/migrate/20190304192641_add_provider_guids.rb]: Also you added the mixin for regions, but are not using `.in_my_region` here ...

**Comment by d-m-u** [comment]
@jrafanie can I please get eyes on this?

**Comment by djberg96** [comment]
Just poking my nose in where it doesn't belong...

Postgres has supported a uuid data type since 8.3. Is that something we could leverage here instead of relying on AR?

**Comment by jrafanie** [comment]
> Postgres has supported a uuid data type since 8.3. Is that something we could leverage here instead of relying on AR?

@djberg96 database agnosticism originally... now?  I'm not sure.  I guess it depends if changing it to uuid buys us anything considering we already chose a different column type.  Maybe?

**Comment by djberg96** [comment]
@jrafanie It would allow for greater flexibility in terms of input formatting, and you could also setup the schema to autogenerate a default value.

https://www.postgresql.org/docs/10/datatype-uuid.html and https://www.postgresql.org/docs/10/uuid-ossp.html

**Comment by d-m-u** [comment]
@miq-bot assign @Fryguy

**Comment by bdunne** [comment]
@d-m-u Was this resolved by another PR?

File: spec/migrations/20190304192641_add_provider_guids_spec.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
+require_migration
+
+describe AddProviderGuids do
+  let(:provider) { migration_stub(:Provider) }
+
+  migration_context :up do
+    it "sets provider guids" do
+      obj1 = provider.create!
+      obj1.save(:validate => false)
```
Comment (bdunne) [inline spec/migrations/20190304192641_add_provider_guids_spec.rb]: Why `:validate => false` here?  Should this have been `dup`'ed?

**Comment by bdunne** [comment]
@carbonin Any other concerns?
